lib/tapctl.ml

```diff
@@ -1,450 +1,451 @@
+        write_dummy_tapdisk_list ctx list)
+
+  let list ?t ctx =
+    Mutex.execute d_lock (fun () ->
+        let list = get_dummy_tapdisk_list ctx in
+        List.filter_map (fun e ->
+            let args =
+              match Opt.map (String.split ':') e.d_args with
+              | Some (ty::arguments) ->
+                Some (ty,String.concat ":" arguments)
+              | _ -> None
+            in
+            match (e.d_minor, e.d_pid, e.d_state, t) with
+            | Some m, Some p, Some s, None ->
+              Some ({tapdisk_pid=p; minor=m},s,args)
+            | Some m, Some p, Some s, Some t ->
+              if t.tapdisk_pid = p && t.minor=m then
+                Some ({tapdisk_pid=p; minor=m},s,args)
+              else
+                None
+            | _ -> None) list)
+
+  let stats ctx t =
+    let open Stats in
+    { name = "none";
+      secs = 0L,0L;
+      images = [];
+      tap = { Tap.minor = t.minor;
+              reqs = 0L,0L;
+              kicks = 0L,0L; };
+      nbd_mirror_failed = 0;
+      reqs_outstanding = 0;
+    }
 end
- 		
+
 
 (* END OF DUMMY STUFF *)
 
 let colon = Re_str.regexp_string ":"
 
 let canonicalise x =
-	if not(Filename.is_relative x)
-	then x
-	else begin
-		(* Search the PATH and XCP_PATH for the executable *)
-		let paths = Re_str.split colon (Sys.getenv "PATH") in
-		let xen_paths = try Re_str.split colon (Sys.getenv "XCP_PATH") with _ -> [] in
-		let first_hit = List.fold_left (fun found path -> match found with
-			| Some hit -> found
-			| None ->
-				let possibility = Filename.concat path x in
-				if Sys.file_exists possibility
-				then Some possibility
-				else None
-		) None (xen_paths @ paths) in
-		match first_hit with
-		| None -> x
-		| Some hit -> hit
-	end
+  if not(Filename.is_relative x)
+  then x
+  else begin
+    (* Search the PATH and XCP_PATH for the executable *)
+    let paths = Re_str.split colon (Sys.getenv "PATH") in
+    let xen_paths = try Re_str.split colon (Sys.getenv "XCP_PATH") with _ -> [] in
+    let first_hit = List.fold_left (fun found path -> match found with
+        | Some hit -> found
+        | None ->
+          let possibility = Filename.concat path x in
+          if Sys.file_exists possibility
+          then Some possibility
+          else None
+      ) None (xen_paths @ paths) in
+    match first_hit with
+    | None -> x
+    | Some hit -> hit
+  end
 
 let tap_ctl = canonicalise "tap-ctl"
 
 let invoke_tap_ctl ctx cmd args =
-	let find x = try [ x ^ "=" ^ (Sys.getenv x) ] with _ -> [] in
-	let env = Array.of_list (find "PATH" @ (find "TAPDISK") @ (find "TAPDISK2")) in
-	let stdout, stderr = execute_command_get_output ~env tap_ctl (cmd::args) in
-	stdout
+  let find x = try [ x ^ "=" ^ (Sys.getenv x) ] with _ -> [] in
+  let env = Array.of_list (find "PATH" @ (find "TAPDISK") @ (find "TAPDISK2")) in
+  let stdout, stderr = execute_command_get_output ~env tap_ctl (cmd::args) in
```

Comment:
I think this could be replaced by `Unix.open_process_in` as in 
```ocaml
let read_process_lines command =
  let in_channel = Unix.open_process_in command in
  let rec fill lines =
    try
        let line = input_line in_channel in
        fill line::lines
    with End_of_file ->
      ignore (Unix.close_process_in in_channel);
      lines
  in
  List.rev lines
```
and drop the dependency on `forkexec` but probably this should be a separate PR

Reply:
It belongs to a different PR :)